File: paa/randomized_select/randomized_select.c
#include<stdio.h>
#include<stdlib.h>

void troca(int *x, int *y);
int  partition(int v[], int p, int r);
int  randomized_partition(int v[], int p, int r);
int  randomized_select(int v[], int p, int r, int i);

void main() {
    int vetor[] = {30,35,20,10,40,50,80,100}, i;
    i = randomized_select(vetor, 0, 7, 3);

    printf("%d \n", i);
}

void troca(int *x, int *y) {
    int temp = *x;
    *x = *y;
    *y = temp;
}

int partition(int vetor[], int posicao_inicial, int posicao_final) {
    int pivot = vetor[posicao_final];
    int i = posicao_inicial-1, j;

    for(j=posicao_inicial; j <= posicao_final-1; j++) {
        if(vetor[j] <= pivot) {
            i++;
            troca(&vetor[i], &vetor[j]);
        }
    }
    troca(&vetor[i+1], &vetor[posicao_final]);
    return i+1;
}

int randomized_partition(int vetor[], int posicao_inicial, int posicao_final) {
    int posicao_aleatoria = rand()/(RAND_MAX/posicao_final);
    troca(&vetor[posicao_final], &vetor[posicao_aleatoria]);
    return partition(vetor, posicao_inicial, posicao_final);
}

int randomized_select(int vetor[], int posicao_inicial, int posicao_final, int i) {
    if(posicao_inicial == posicao_final)
        return vetor[posicao_inicial];

    int q = randomized_partition(vetor, posicao_inicial, posicao_final);
    int k = q - posicao_inicial + 1;

    if(i == k)  //o pivo eh a resposta
        return vetor[q];
    else if (i < k)
        return randomized_select(vetor, posicao_inicial, posicao_final-1, i);
    else
        return randomized_select(vetor, q+1, posicao_final, i-k);
}

File: paa/randomized_select/randomized_select.py
#encoding: utf-8
import random

def RANDOMIZED_PARTITION(vetor, p, r):
    ultima_posicao = r-1
    print "- Randomized Partition comecou:",  vetor, "Verificando de ", p,"ate", r
    pivot = vetor[ultima_posicao]
#    pivot = random.randrange(p, r)

    print "- Pivot escolhido", pivot
    i = p
    for j in range(p, ultima_posicao):
        if vetor[j] <= pivot:
            vetor[i], vetor[j] = vetor[j], vetor[i]
            i += 1
    vetor[i], vetor[ultima_posicao] = vetor[ultima_posicao], vetor[i]
    print "- Randomized Partition terminou:", vetor, " Retornando:", i
    return i

def RANDOMIZED_SELECT(vetor, posicao_inicial, posicao_final, i):
    print "RANDOMIZED_SELECT", vetor, posicao_inicial, posicao_final, i
    if posicao_final + 1 == posicao_inicial:
        return vetor[posicao_inicial]
    q = RANDOMIZED_PARTITION(vetor, posicao_inicial, posicao_final)
    k = q - posicao_inicial
    if i == k:
        print "O escolhido é o pivot"
        return vetor[q]
    elif i < k:
        print "[i %d < k %d]" %(i,k), vetor, posicao_inicial, q, i
        return RANDOMIZED_SELECT(vetor, posicao_inicial, q, i)

    else:
        print "[i %d > k %d]" %(i,k), vetor, q, posicao_final, i-k
        return RANDOMIZED_SELECT(vetor, q, posicao_final, i - k)


